File: 2020/src/problem_three.py
from collections import defaultdict


def main():
    f = open("s3.4-31.in")
    parsed = f.read().splitlines()
    needle = parsed[0]
    haystack = parsed[1]

    freqs = init_freqs(needle)

    eval(needle, haystack, freqs)


def init_freqs(needle):
    hashmap = {}

    for i in range(len(needle)):
        if needle[i] not in hashmap:
            hashmap[needle[i]] = 1
        else:
            hashmap[needle[i]] += 1

    return hashmap


def eval(needle, haystack, needle_freqs):
    needle_len, haystack_len = len(needle)-1, len(haystack)

    perms = 0
    prev_perms = {}
    prev_chars_freqs = defaultdict(default_freq_value)

    for i in range(haystack_len):
        perm_found = True
        prev_chars_freqs[haystack[i]] += 1

        if not i >= needle_len:
            continue

        for key in needle_freqs:
            if needle_freqs[key] == prev_chars_freqs[key]:
                continue
            perm_found = False
        
        section = haystack[i-needle_len:i+1]
        if perm_found and section not in prev_perms:
            prev_perms[section] = True
            perms += 1

        if prev_chars_freqs[haystack[i-needle_len]] > 0:
            prev_chars_freqs[haystack[i-needle_len]] -= 1

    print(perms)


def default_freq_value():
    return 0


main()

File: 2020/src/problem_three.cc
using namespace::std;
#include <iostream>
#include <string>
#include <map>


map<char, int> init_needle_freqs(string needle)
{
    map<char, int> hashmap;

    for (int i = 0; i < needle.length(); i++)
    {
        char curr_char = needle[i];

        hashmap[curr_char]++;

        if (hashmap[curr_char] == 0)
            hashmap[curr_char]++;
    }

    return hashmap;
}

void eval(string needle, string haystack, map<char, int> needle_freqs)
{
    int needle_len, haystack_len, perms;
    map<string, bool> prev_perms;
    map<char, int> prev_char_freqs;

    perms = 0;
    needle_len = needle.length()-1;
    haystack_len = haystack.length()-1;

    for (int i = 0; i < haystack_len+1; i++)
    {
        bool perm_found = true;
        prev_char_freqs[haystack[i]]++;
        if (prev_char_freqs[haystack[i]] == 0)
            prev_char_freqs[haystack[i]]++;

        if (!(i >= needle_len))
            continue;

        for (auto const& [key, val] : prev_char_freqs)
        {
            if (prev_char_freqs.count(key) > 0 && prev_char_freqs[key] == needle_freqs[key])
                continue;
            perm_found = false;
        }

        string section = haystack.substr(i-needle_len, needle_len+1);
        if (perm_found && prev_perms.count(section) == 0)
        {
            prev_perms[section] = true;
            perms += 1;
        }
        
        if (prev_char_freqs[haystack[i-needle_len]] > 0)
            prev_char_freqs[haystack[i-needle_len]]--;
    }

    cout << perms;
}


int main()
{
    string needle, haystack;

    cin >> needle;
    cin >> haystack;

    map<char, int> needle_freqs = init_needle_freqs(needle);

    eval(needle, haystack, needle_freqs);
    
    return 0;
}


